### ISSUE TITLE
alacritty becomes unresponsive on Openbox/X11 intel HD520

### ISSUE THREAD

**Issue by alexyuriev** [alacritty/alacritty#3284]
From time to time, alacritty becomes unresponsive - keys do not work, screen freezes. This can be in the middle of something sending an output to display ( such as cat largefile.txt ) or system being completely idle. I cannot reliably reproduce it but in my normal workflow it happens between 5-15 minutes. When alacritty freezes key presses are swallowed.

To unfreeze, I use hotkeys used by the openbox to switch from one workspace to another and back. These hotkeys are not passed to processed by openbox and are not passed to alacritty at all (by openbox design) Upon returning the workspace with the previously frozen alacritty window, the swallowed keypresses are properly displayed on screen.

Note: I have two nearly identical systems with identical configurations of openbox and alacritty. The only system exhibiting this behavior is Intel HD520. Nvidia (prop X11 ) does not exhibit this issue. 

I'm  happy to provide help debugging this.

### System

OS: Linux
Version: alacritty 0.4.1
Linux/BSD: X11, Openbox, Intel HD520

**Comment by kchibisov** [comment]
Could you try the latest master?

**Comment by alexyuriev** [comment]
I will try tomorrow.

**Comment by alexyuriev** [comment]
Result with 0.4.2 - still no freezes on NVidia, and fewer ( though non-scientific observation ) but still about once every ~20-30 minutes on Intel HD520.

I have installed it now on my main laptop, so I will push it hard today and see if I can figure out any patterns.

**Comment by kchibisov** [comment]
Btw, does `ctrl + s` in your shell do something? I'm asking since I've thought that you're pressing it accidentaly due to bug on v0.4.1, since just pressing `s` was able to trigger in some cases(without control).

if `ctrl + s` works, use `ctrl + q` to exit it.

**Comment by chrisduerr** [comment]
Also if that's not what you're experiencing, I'd recommend checking driver vsync settings. Since Alacritty renders on vsync, it's possible for your GPU driver to block us indefinitely.

**Comment by alexyuriev** [comment]
> Btw, does `ctrl + s` in your shell do something? I'm asking since I've thought that you're pressing it accidentaly due to bug on v0.4.1, since just pressing `s` was able to trigger in some cases(without control).
> 
> if `ctrl + s` works, use `ctrl + q` to exit it.

That was my first thought, but `ctrl + q` did not fix it, as it does after `ctrl + s` so it is not that. 

I will look at vsync next.

**Comment by alexyuriev** [comment]
Ok, so here's an update from today:

I cannot still reliably replicate the issue *but* it definitely involves interaction with a mouse! If I am very careful to neither use a touch pad nor an external mouse, I do not experience any lockups.

**Comment by chrisduerr** [comment]
Seems like this might be a duplicate of #2453 then? Assuming that it really only happens on copy/paste and not with normal mouse interaction.

**Comment by alexyuriev** [comment]
I will play with backtracing it today. It is not Wayland tho. Plain X11

**Comment by alexyuriev** [comment]
Any thoughts on how to generate a useful debug log for this?

**Comment by chrisduerr** [comment]
I'm not aware of any log output that would help here.

**Comment by alexyuriev** [comment]
@chrisduerr Does alacritty have any global timers that are firing that maybe I can hook up some rudimentary logic to see if the timers are hung when this behavior occurs? If so where would I find them?

**Comment by chrisduerr** [comment]
Alacritty does not have any timers in it at all actually, which is why blinking cursors and scrolling with selection do not work yet. Except for timers than are implicitly used by other libraries of course, like live config reload might when the system doesn't support notifications about file changes.

I think the most useful output you're going to get is the `alacritty --print-events -vvv` output with the latest master. I just changed our log output to include nanosecond precision, instead of just minutes, so it should be possible to see when nothing happened for a few seconds. At least if you can narrow down the exact time it happened in the logfile it might be worth inspecting that section.

**Comment by alexyuriev** [comment]
Ok, I'm going through a heap of logs now and hopefully i would be able to isolate something.

**Comment by kchibisov** [comment]
I'd recommend you to wait until it freezes and attach to it with gdb afterwards, so you can try to find where it locks. Also you should build it with debug symbols (our default release build).

**Comment by neworld** [comment]
I have a very similar issue but in addition Spotify, visual code and chrome freeze at the same time. Just chrome somehow is able to unfreeze itself. It looks like GPU related problem.

**Comment by alexyuriev** [comment]
I'm starting to believe this might be Intel HD520 related. I have been going through the different Linux distributions and I'm unable to create a reasonable test case where I can reliably trigger it. 

Does anyone have it happen with HD520 or other Intel GPUs?

**Comment by mbaz** [comment]
I use openbox on X11, on an Ice Lake CPU (UHD). I haven't seen the problem you describe (I've owned this laptop since November 2019).

**Comment by neworld** [comment]
I have a similar problem with UHD  630:
```
    Vendor: Intel Open Source Technology Center (0x8086)
    Device: Mesa DRI Intel(R) UHD Graphics 630 (Coffeelake 3x8 GT2)  (0x3e9b)
    Version: 19.3.4
    Accelerated: yes
    Video memory: 3072MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.6
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
```

**Comment by alexyuriev** [comment]
@neworld 

I'm also running UHD (4K). This is mine:

```
    Vendor: Intel Open Source Technology Center (0x8086)
    Device: Mesa DRI Intel(R) HD Graphics 520 (Skylake GT2)  (0x1916)
    Version: 18.3.6
    Accelerated: yes
    Video memory: 3072MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
```

@mbaz Could you post yours?

**Comment by mbaz** [comment]
@alexyuriev What command do I run to get that information?

**Comment by neworld** [comment]
`glxinfo -B`

**Comment by mbaz** [comment]
Thanks!
```
    Vendor: Intel Open Source Technology Center (0x8086)
    Device: Mesa DRI Intel(R) UHD Graphics (Ice Lake 4x8 GT1)  (0x8a56)
    Version: 19.3.4
    Accelerated: yes
    Video memory: 3072MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.6
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
```

**Comment by Inc0ns1st3nt** [comment]
I have a similar issue, and I ve got a UHD 620 (I think) graphics as well. 
For me, alacritty fires up, and it's just straight up being unresponsive, only pressing enter works, but that just refresh alacritty to be at prompt again.
I am also having this issue for kitty, discord.

Should we install nvidia package or something?

**Comment by kchibisov** [comment]
I feel like this is a graphics driver bug tbh, like GPU vendor + driver(i915 in kernel) is the same (intel) for all the users here.

**Comment by Inc0ns1st3nt** [comment]
It seems that the app redraws itself after a fullscreen shortcut on my setup,
interestingly my friend who has an UHD 620, i915 driver, it works for him.

**Comment by kchibisov** [comment]
I mean it depends on a kernel + some other stuff, for me it also works, however I've seen weird things on my machine with intel GPU recently (like micro freeze). I've heard that there're some problems with newer kernels like 5.2+? IIRC, but I'm not sure about this information. Also, mesa changed intel driver to IRIS recently instead of i965 for some new GPUs, like there's a lot of things that could be related tbh.

**Comment by Inc0ns1st3nt** [comment]
smh, I have found some info in particular related to i915 and intel graphics card, I am not sure if this could be related: [OpenGL_2.1_with_i915_driver](
https://wiki.archlinux.org/index.php/Intel_graphics#OpenGL_2.1_with_i915_driver)

**Comment by kchibisov** [comment]
That's unrelated, since it's talking about old mesa.

**Comment by mwip** [comment]
I experience the same behavior using UHD 620 but not on my Radeon HD 6990. I'm on Xmonad. Once alacritty instance freezes, It is just an empty frame for the WM with wierd glitches much like a frozen Win98 Window.
Here's my glxinfo: 
```
    Vendor: Intel (0x8086)
    Device: Mesa Intel(R) UHD Graphics 620 (WHL GT2) (0x3ea0)
    Version: 20.0.4
    Accelerated: yes
    Video memory: 3072MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.6
    Max compat profile version: 4.6
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
```

I'm running `alacritty 0.4.2` from the Arch repos.

**Comment by mwip** [comment]
Using `alacritty --print-events` I found, that when alacritty freezes, it does not receive any events anymore.

**Comment by kchibisov** [comment]
Could someone with gdb show the place where it freezes? It could help a lot to understand your issue. You can attach to a running process with `sudo gdb -p PID` and just to `bt full`/`info threads`/other useful  tools, etc, like it could help a lot, since it should show the point of freeze. Also, compile alacritty with debug symbols before doing so. Our default build is with debug symbols though, so just running `cargo build --release` already is a good binary for debugging.

**Comment by neworld** [comment]
Unfortunately archlinux ships alacrity without symbols. But at least I know drivers are to blame. Maybe it could help?
```
#0  0x00007f5b5a46eabf in poll () from /usr/lib/libc.so.6
#1  0x00007f5b5a59f63b in ?? () from /usr/lib/libxcb.so.1
#2  0x00007f5b5a5a145b in xcb_wait_for_special_event () from /usr/lib/libxcb.so.1
#3  0x00007f5b57059931 in glPrimitiveBoundingBox () from /usr/lib/libGLX_mesa.so.0
#4  0x00007f5b57059a98 in glPrimitiveBoundingBox () from /usr/lib/libGLX_mesa.so.0
#5  0x00007f5b5705acce in glPrimitiveBoundingBox () from /usr/lib/libGLX_mesa.so.0
#6  0x00007f5b5705bc3c in glPrimitiveBoundingBox () from /usr/lib/libGLX_mesa.so.0
#7  0x00007f5b563d2f3a in ?? () from /usr/lib/dri/i965_dri.so
#8  0x00007f5b563d36e5 in ?? () from /usr/lib/dri/i965_dri.so
#9  0x00007f5b563cf155 in ?? () from /usr/lib/dri/i965_dri.so
```

**Comment by kchibisov** [comment]
It points out that the bug isn't in alacritty.

**Comment by neworld** [comment]
Reported the issue [to mesa](https://gitlab.freedesktop.org/mesa/mesa/-/issues/2744).

**Comment by Inc0ns1st3nt** [comment]
I ran `info threads`:
```
  Id   Target Id                                           Frame 
* 1    Thread 0x7faebc60c940 (LWP 349527) "alacritty"      0x00007faebcbc870e in epoll_wait () from /usr/lib/libc.so.6
  2    Thread 0x7faeb320f700 (LWP 349543) "alacrit:disk$0" 0x00007faebccb8cf5 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  3    Thread 0x7faeb2a0e700 (LWP 349544) "alacrit:disk$1" 0x00007faebccb8cf5 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  4    Thread 0x7faeb220d700 (LWP 349545) "alacrit:disk$2" 0x00007faebccb8cf5 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  5    Thread 0x7faeb1a0c700 (LWP 349546) "alacrit:disk$3" 0x00007faebccb8cf5 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
  6    Thread 0x7faeb044e700 (LWP 349547) "alacritty"      0x00007faebcbbdabf in poll () from /usr/lib/libc.so.6
  7    Thread 0x7faeb024b700 (LWP 349548) "alacritty"      0x00007faebcbbdabf in poll () from /usr/lib/libc.so.6
  8    Thread 0x7faeabfff700 (LWP 349550) "pty reader"     0x00007faebcbc870e in epoll_wait () from /usr/lib/libc.so.6
```
Now `bt full`:
```
#0  0x00007faebcbc870e in epoll_wait () from /usr/lib/libc.so.6 No symbol table info available.
#1  0x00005639b2bb55f9 in ?? ()                                              No symbol table info available.
#2  0x00005639b29baeea in ?? ()                                             No symbol table info available.
#3  0x00005639b2afbc7e in ?? ()                                              No symbol table info available.
#4  0x00005639b2ad1753 in ?? ()                                             No symbol table info available.
#5  0x00005639b2af2aa5 in ?? ()                                              No symbol table info available.
#6  0x00007faebcaf0023 in __libc_start_main () from /usr/lib/libc.so.6 No symbol table info available.
#7  0x00005639b29690de in ?? ()                                             No symbol table info available.
```
My issue seems to be slightly different with some people here.
> I have a similar issue, and I ve got a UHD 620 (I think) graphics as well.
> For me, alacritty fires up, and it's just straight up being unresponsive, only pressing enter works, but that just refresh alacritty to be at prompt again.
> I am also having this issue for kitty, discord.
Bit more detail:
> It seems that the app redraws itself after a fullscreen shortcut on my setup,

**Comment by neworld** [comment]
I compiled `mesa` with debug flag and symbols, and my system became stable. Only one freeze in two days. And stacktrace a far different. I suppose the last stacktrace was bad of some gdb. Now I found:

```
#0  0x00007ffff7d00abf in poll () from /usr/lib/libc.so.6
#1  0x00007ffff7e5063b in ?? () from /usr/lib/libxcb.so.1
#2  0x00007ffff7e5245b in xcb_wait_for_special_event () from /usr/lib/libxcb.so.1
#3  0x00007ffff48f39f1 in dri3_wait_for_event_locked (draw=0x555556c932a8) at ../mesa-20.0.4/src/loader/loader_dri3_helper.c:545
.....
#14 0x00007ffff32a16a4 in st_Clear (ctx=0x555556c64c40, mask=2) at ../mesa-20.0.4/src/mesa/state_tracker/st_cb_clear.c:438
#15 0x0000555555854537 in alacritty::gl::Clear (mask=16384) at /home/neworld/Downloads/alacritty-0.4.2/target/debug/build/alacritty-b9da0f4f450cace8/out/gl_bindings.rs:1604
....
```

Does it mean the problem is on the X11 side?

Stacktraces:

[mesa.full.txt](https://github.com/alacritty/alacritty/files/4457752/mesa.full.txt)
[mesa.txt](https://github.com/alacritty/alacritty/files/4457753/mesa.txt)

**Comment by kchibisov** [comment]
if it blocked on `glClear` it doesn't look really good tbh.

**Comment by neworld** [comment]
I found this issue is quite popular. For example, the one was reported for kitty: https://github.com/kovidgoyal/kitty/issues/1483. People suggest downgrade drivers:

> I have also run into this, I think that downgrading xf86-video-intel to a version from January fixes it for now

**Comment by chrisduerr** [comment]
Sounds like it's a driver bug then, not because of Alacritty. So I recommend following the mesa issue.
